### PR TITLE
[DOCS] Reaction Guide

### PR DESCRIPTION
**Pull request by ImVaskel**
## Summary

Adds a reaction guide

## Checklist

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: docs/guide/topics/reactions.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
+.. currentmodule:: discord
+
+.. _guide_topic_reactions
+
+Reactions
+==========
+
+Reactions are a feature within Discord where you can add emojis next to messages.
+They have many uses, but a common one for bots is for polls or giving feedback to a command without sending a new message.
+Anyone can react to a message.
```
Comment (slice) [inline docs/guide/topics/reactions.rst:10]:
```suggestion
Members with the :attr:`~Permissions.add_reactions` can add new reactions to messages. However, this permission is not necessary to add to an existing reaction.
```

The note on line 31 also gets this across, but I feel like it conflicts with this sentence as it is.

File: docs/guide/topics/reactions.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    :scale: 50%
+
+
+.. note::
+
+    You must have the permission :attr:`~Permissions.read_message_history` to add reactions.
+    Additionally, if no one else has reacted to the message using this emoji, you will also need the
+    :attr:`~Permissions.add_reactions` permission.
+
+:meth:`~Message.add_reaction` can also take many types, not just :class:`str`. This includes :class:`~Emoji` or :class:`~PartialEmoji`.
```
Comment (slice) [inline docs/guide/topics/reactions.rst:37]:
```suggestion
In addition to :class:`str`, you can also pass :class:`~Emoji`, :class:`~PartialEmoji`, or :class:`~Reaction` to :meth:`~Message.add_reaction`.
```

Exhaustively listing everything that `add_reaction` can accept here might help.

File: docs/guide/topics/reactions.rst [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+For example, the following shows an alternative way to add a custom emoji reaction:
+
+.. code-block:: python3
+
+    emoji = client.get_emoji(819046807370989599)
+    await message.add_reaction(emoji)
+
+You may notice that if you try to add a standard Discord emoji like ``:smile:``, you'll receive an error.
+This is because to add standard Discord emojis to a message, you must pass their unicode values.
+Because of this, to add 😄 / ``:smile:`` you would do either of the following:
```
Comment (slice) [inline docs/guide/topics/reactions.rst:47]:
```suggestion
When adding standard (i.e. not custom) emojis as a reaction, you cannot pass emoji shortcodes,
such as the string ``":smile:"`` in order to react with 😄. The Discord client recognizes these
shortcodes as a typing convenience, but the API is not aware of them. So, in your bot, you must
pass the actual Unicode codepoint as is (e.g., ``"😄"``). This can be done like so:
```

This seems a bit pedantic, but the nomenclature can be ambiguous. "Emoji" can mean actual Unicode codepoints, Discord's custom emoji, or non-standard emoji shortcodes. I tried to reword stuff to make it less confusing here.

File: docs/guide/topics/reactions.rst [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+.. code-block:: python3
+
+    # pass the unicode emoji via the hexadecimal value
+    await message.add_reaction("\U0001f604")
```
Comment (slice) [inline docs/guide/topics/reactions.rst:52]:
```suggestion
    # Pass the emoji codepoint via 32-bit hexadecimal escape.
    # For more information, see: https://docs.python.org/3/howto/unicode.html#the-string-type
    await message.add_reaction("\U0001f604")
```

File: docs/guide/topics/reactions.rst [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    
+    # or
+
+    # get the unicode emoji via the name (obtained via the unicode database)
```
Comment (slice) [inline docs/guide/topics/reactions.rst:56]:
```suggestion
    # Alternatively, use the codepoint's name.
    # See: https://docs.python.org/3/library/unicodedata.html
```

File: docs/guide/topics/reactions.rst [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    users = await reaction.users().flatten()
+    # users is now a list of User...
+    winner = random.choice(users)
+    await channel.send(f'{winner} has won the raffle.')
+
+Next Steps
+~~~~~~~~~~~
+
+You can find further information about :class:`~Reaction` in its documentation.
+Additionally, :meth:`~Client.wait_for` can pair well with reactions.
```
Comment (slice) [inline docs/guide/topics/reactions.rst:133]: Might wanna link to the `wait_for` guide here? It doesn't exist yet though.

Comment (ImVaskel) [reply docs/guide/topics/reactions.rst:133]: i will link it when the guide exists, but currently it does not.

File: docs/guide/topics/reactions.rst (hunk 3 above)
Comment (Vexs) [reply docs/guide/topics/reactions.rst:47]: `:name:` emojis are supported in other contexts as well- such as sending messages.

Comment (slice) [reply docs/guide/topics/reactions.rst:47]: Unless I'm missing something, shortcodes in messages don't seem to be postprocessed by the API in any way, so the remark "[...] the API is not aware of them" appears to hold:

image

The client parses them though, of course.

File: docs/guide/topics/reactions.rst [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+.. image:: /images/guide/topics/reactions/remove_reaction_after.png
+    :scale: 50%
+
+Removing all reactions of a single emoji from a message is also possible: 
+
+.. code-block:: python3
+
+    await message.clear_reaction("<:amogus:819046807370989599>")
+
+It is also possible to remove all reactions from a message, not just a singular member or emoji:
```
Comment (Gorialis) [inline docs/guide/topics/reactions.rst:94]: This following sentence feels a touch staggered due to the previous sentence (e.g., saying "You can also x. You can also y. You can also z." with no bridges doesn't read very well).

I think it'd probably be best to fix it like:
```suggestion
Removing all reactions of a given emoji from a message is also possible: 

.. code-block:: python3

    await message.clear_reaction("<:amogus:819046807370989599>")

And we can also remove all reactions entirely:
```

File: docs/guide/topics/reactions.rst (hunk 1 above)
Comment (ImVaskel) [reply docs/guide/topics/reactions.rst:10]: i feel like a note would be better than adding that, perhaps just removing the "anyone can react to a message"?

File: docs/guide/topics/reactions.rst (hunk 3 above)
Comment (ImVaskel) [reply docs/guide/topics/reactions.rst:47]: i do not see a reason to introduce client semantics to the guide.

Comment (slice) [reply docs/guide/topics/reactions.rst:47]: I feel like reducing the amount of potential confusion is desirable here, even at the expense of the result being a bit too wordy or technical. I don't think it's particularly difficult or overtly technical to explain that the API does not recognize shortcodes once it's spelled out, and that emojis are semantically distinct from shortcodes in this way.

I do see that this can be over the top here, though. Oh well pensive

Comment (ImVaskel) [reply docs/guide/topics/reactions.rst:47]: i feel like maybe a "the client and api see emojis in different ways, so you should be doing "X" to make sure you appease the api (in other wording of course).

Comment (slice) [reply docs/guide/topics/reactions.rst:47]: Oh, that could help, yeah. That's basically what you had in the first place, though, I think. So it's fine without my suggestion, seems.

File: docs/guide/topics/reactions.rst (hunk 1 above)
Comment (slice) [reply docs/guide/topics/reactions.rst:10]: > perhaps just removing the "anyone can react to a message"?

Yeah that seems good.

File: docs/guide/topics/reactions.rst (hunk 6 above)
Comment (AbstractUmbra) [reply docs/guide/topics/reactions.rst:133]: The guide now exists and has a reference link, this can be added.

**Review by AbstractUmbra — COMMENTED** [review]
Only one comment here.

File: docs/guide/topics/reactions.rst [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    :scale: 50%
+
+You can find the documentation for adding reactions to a message object at :meth:`~Message.add_reaction`.
+
+Removing a reaction
+~~~~~~~~~~~~~~~~~~~~~
+
+Removing reactions is similar to adding a reaction, but you must also specify a member to remove the reaction from.
+This can be anything that meets the abc :class:`~abc.Snowflake`, like :class:`~discord.Member` or :class:`~discord.Object`.
+This means that you cannot pass simply only an id, and you must instead pass something with an ``id`` attribute.
```
Comment (numbermaniac) [inline docs/guide/topics/reactions.rst:71]:
```suggestion
This means that you cannot simply pass only an id, you must instead pass an object with an ``id`` attribute.
```

IMO sounds a little bit more natural this way.

**Review by Soheab — CHANGES_REQUESTED** [review]
This should cover super/burst reactions too.
